### PR TITLE
Added initial CustomOp template generator

### PR DESCRIPTION
**Pull request by sayanshaw24**
Creates `custom_op_template.hpp` file in `tools` directory when run with ONNX model.

Usage: `python tools\gen_customop_template.py onnx_model_filepath`

### CODE REVIEW

File: tools/gen_customop_template.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+# Load and check ONNX model
+model = onnx.load(sys.argv[1])
+onnx.checker.check_model(model)
+
+# Get input and output type count from the ONNX model using protobuf information
+input_count = 0
+output_count = 0
+
+# Note: if there is no custom op node in the graph, GetInputTypeCount() and
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply tools/gen_customop_template.py]: Yep, thanks! Added that.

File: tools/gen_customop_template.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+Generates the C++ template code for a new custom operator.
+
+Usage: $ python tools\gen_customop_template.py onnx_model_filepath
+'''
+
+import onnx
+import sys
+
+# Load and check ONNX model
+model = onnx.load(sys.argv[1])
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply tools/gen_customop_template.py]: Sounds good!

File: tools/gen_customop_template.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+'''
+Generates the C++ template code for a new custom operator.
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply tools/gen_customop_template.py:2]: Will do.

File: tools/gen_customop_template.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    custom_op_node_exists = False
+
+    for node in model.graph.node:
+        # Find CustomOp node using domain
+        if node.domain == "ai.onnx.contrib" or node.domain == "com.microsoft.extensions":
+            custom_op_node_exists = True
+            input_count = len(node.input)
+            output_count = len(node.output)
+
+    if not custom_op_node_exists:
```
Comment (edgchen1) [inline tools/gen_customop_template.py:31]: what if there is more than one?

Comment (sayanshaw24) [reply tools/gen_customop_template.py:31]: From my conversation with @wenbingl I believe we are just handling one custom op node case as of now - this is just an initial wip, there is also more research to be done on shape inference to support the template more extensively.

File: tools/gen_customop_template.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    // mapping of TensorProto.DataType value
+    ONNXTensorElementDataType {custom_op}::GetOutputType(size_t /*index*/) const {{
+      return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+    }};
+    '''
+
+    hpp_file = customop_template.format(custom_op = op, op_name = op_name, input_type_count = input_type_count, output_type_count = output_type_count)
+
+    # Write code to C++ temmplate file
+    new_op = "tools/custom_op_template.hpp"
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply tools/gen_customop_template.py:76]: yep! i'll change it

File: tools/gen_customop_template.py (hunk 4 above)
Comment (wenbingl) [reply tools/gen_customop_template.py:31]: Generally, either you need to handle the case, or print a message to let the user know.

File: tools/gen_customop_template.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+import argparse
+import onnx
+import pathlib
+import sys
+
+# Load and check ONNX model
+def load_onnx_model(model_path):
+    model = onnx.load(model_path)
+    onnx.checker.check_model(model)
```
Comment (wenbingl) [inline tools/gen_customop_template.py:15]: Did you verify that the model check passes if there is a custom operator in the model?

**Review by wenbingl — COMMENTED** [review]
Can you also add some examples in tutorial to demonstrate how it works in end to end?

File: tools/gen_customop_template.py (hunk 6 above)
Comment (sayanshaw24) [reply tools/gen_customop_template.py:15]: Yep! Will add a test for the same.

File: tools/gen_customop_template.py (hunk 4 above)
Comment (sayanshaw24) [reply tools/gen_customop_template.py:31]: Sounds good!

File: tools/gen_customop_template.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    ONNXTensorElementDataType {custom_op}::GetOutputType(size_t /*index*/) const {{
+      return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+    }};
+    '''
+
+    hpp_file = customop_template.format(custom_op = op, op_name = op_name, input_type_count = input_type_count, output_type_count = output_type_count)
+
+    # Write code to C++ template file
+    with open(customop_template_filepath,'wt') as file:
+        print("Added C++ CustomOp template code to output filepath: " + str(customop_template_filepath) + "\n")
```
Comment (edgchen1) [inline tools/gen_customop_template.py:77]: print() will add a newline at the end by default.
f-strings are a convenient shorthand (https://docs.python.org/3/reference/lexical_analysis.html#f-strings):
```suggestion
        print(f"Added C++ CustomOp template code to output filepath: {customop_template_filepath}")
```

File: tools/gen_customop_template.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+def get_io_count(model):
+    input_count = 0
+    output_count = 0
+    custom_op_node_exists = False
+
+    print("Note: This C++ CustomOp template generator currently only supports models with one custom op node.\n________\n")
+    for node in model.graph.node:
+        # Find CustomOp node using domain
+        if node.domain == "ai.onnx.contrib" or node.domain == "com.microsoft.extensions":
+            custom_op_node_exists = True
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply tools/gen_customop_template.py:27]: this assert statement could also be triggered if there were no custom op nodes either right? but sounds good, will change it in similar fashion and add a note mentioning future support for multiple custom op nodes.

File: test/test_tools_customop_template.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    # check input and output type count of models extracted by template generator
+    def check_io_count(self, model_name, output_path, expected_input_count, expected_output_count):
+        model_path = os.path.join(test_data_dir, model_name)
+        input_count, output_count = gen_customop_template.main([model_path, output_path])
+        self.assertEqual(input_count, expected_input_count)
+        self.assertEqual(output_count, expected_output_count)
+
+    def test_template(self):
+        gpt2tok_template_output_path = os.path.join(tools_dir, "custom_op_template_gpt2tok.hpp")
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply test/test_tools_customop_template.py:27]: right, will look into it.

File: tools/gen_customop_template.py (hunk 7 above)
Comment (sayanshaw24) [reply tools/gen_customop_template.py:77]: right, missed replacing that with a f-string, thanks!

Comment (wenbingl) [reply tools/gen_customop_template.py:77]: given ORT stops support of Python < 3.6, we can switch to f-string now

File: tools/gen_customop_template.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+def main(args):
+    args = parse_args(args)
+    model = load_onnx_model(args.onnx_model_filepath)
+    input_count, output_count = get_io_count(model)
+    create_hpp(customop_template_filepath = args.customop_template_filepath, op = "CustomOp", op_name = "CustomOpName", input_type_count = input_count, output_type_count = output_count)
+    return input_count, output_count
+
+if __name__ == "__main__":
+    import sys
```
Comment (wenbingl) [inline tools/gen_customop_template.py:109]: Did you use any Python lint tool to check your source code, like the flake8? https://flake8.pycqa.org/en/latest/.

The tool should suggest not import any module in the middle of source file.

File: test/test_tools_customop_template.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        input_count, output_count = gen_customop_template.main([model_path, output_path])
+        self.assertEqual(input_count, expected_input_count)
+        self.assertEqual(output_count, expected_output_count)
+
+    def test_template(self):
+        template_output_path = os.path.join(test_data_dir, "generated")
+        os.mkdir(template_output_path)
+
+        gpt2tok_template_output_path = os.path.join(template_output_path, "custom_op_template_gpt2tok.hpp")
+        self.check_io_count(model_name = "custom_op_gpt2tok.onnx", output_path = gpt2tok_template_output_path, expected_input_count = 1, expected_output_count = 2)
```
Comment (wenbingl) [inline test/test_tools_customop_template.py:31]: Where did you get these two ONNX files? If they are generated by some Python script, we may just add the generation script in this test file here.

File: tools/gen_customop_template.py (hunk 10 above)
Comment (sayanshaw24) [reply tools/gen_customop_template.py:109]: Got it, will change that.

File: test/test_tools_customop_template.py (hunk 11 above)
Comment (sayanshaw24) [reply test/test_tools_customop_template.py:31]: Sounds good.

**Review by wenbingl — APPROVED** [review]

**Pull request merged**